File: 79-word-search/word-search.java
class Solution {
    boolean[][] used;
    int m;
    int n;
    public boolean exist(char[][] board, String word) {
        m = board.length;
        n = board[0].length;
        used = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (helper(board, word, i, j, 0))
                    return true;
            }
        }
        return false;
    }

    public boolean helper(char[][] board, String word, int i, int j, int idx) {
        if (i < 0 || j < 0|| i >= m || j >= n || used[i][j] || board[i][j] != word.charAt(idx)) return false;
        if (idx == word.length() - 1) return true;
        boolean flag = false;
        used[i][j] = true;
        
            flag = helper(board, word, i + 1, j, idx + 1) ||
                    helper(board, word, i, j + 1, idx + 1) || 
                    helper(board, word, i - 1, j, idx + 1) ||
                    helper(board, word, i, j - 1, idx + 1);
        
        used[i][j] = false;
        return flag;
    }
}

File: 79-word-search/word-search.py
class Solution:
    directions = [
        [0,1], [0,-1],[1,0],[-1,0]
    ]
    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = set()
        m = len(board)
        n = len(board[0])
        wrds = list(word)
        for i in range(0,m):
            for j in range(0,n):
                if self.dfs(board, i, j, wrds, 0, m, n, visited):
                    return True
        return False
    def dfs (self, board, i, j, wrds, start,m,n, visited):
        if start == len(wrds):
            return True
        if 0 <= i < m and 0<= j < n and board[i][j] == wrds[start] and (i,j) not in visited:
            visited.add((i,j))
            for x,y in self.directions:
                if self.dfs(board, i+x, j+y, wrds, start + 1, m, n, visited):
                    return True
            visited.remove((i,j))
        return False